pennylane/ops/qubit.py

```diff
@@ -2051,6 +2082,113 @@ def __init__(self, control_wires=None, wires=None, control_values=None, do_queue
             do_queue=do_queue,
         )
 
+    # pylint: disable=unused-argument
+    def decomposition(self, *args, **kwargs):
+
+        if len(self.control_wires) > 2 and len(self._work_wires) == 0:
+            raise ValueError(f"At least one work wire is required to decompose operation: {self}")
+
+        flips1 = [
+            qml.PauliX(self.control_wires[i])
+            for i, val in enumerate(self.control_values)
+            if val == "0"
+        ]
+
+        if len(self.control_wires) == 1:
+            decomp = [qml.CNOT(wires=[self.control_wires[0], self._target_wire])]
+        elif len(self.control_wires) == 2:
+            decomp = [qml.Toffoli(wires=[*self.control_wires, self._target_wire])]
+        else:
+            num_work_wires_needed = len(self.control_wires) - 2
+
+            if len(self._work_wires) >= num_work_wires_needed:
+                decomp = self._decomposition_with_many_workers(
+                    self.control_wires, self._target_wire, self._work_wires
+                )
+            else:
+                work_wire = self._work_wires[0]
+                decomp = self._decomposition_with_one_worker(
+                    self.control_wires, self._target_wire, work_wire
+                )
+
+        flips2 = [
+            qml.PauliX(self.control_wires[i])
+            for i, val in enumerate(self.control_values)
+            if val == "0"
+        ]
+
+        return flips1 + decomp + flips2
+
+    @staticmethod
+    def _decomposition_with_many_workers(control_wires, target_wire, work_wires):
```

Comment:
My thoughts are that since `w0`,`w1`, and `w2` are work qubits that start in `|0>` these CCX gates in the beginning will not have an effect on `t1` or the work qubits. Then we wouldn't need to uncompute them at the end either.

Reply:
I hadn't realized that the work wires can be in a non-|0> state and still give the right results! That's exciting resource-wise!